### DISCUSSION TITLE
GeneratedCodeAttribute cannot be inherited

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/roslyn#65433] [Q&A] [answered]
My source generator attaches the `GeneratedCode` attribute to the generated code (class).
This works fine until I have to to derive the generated class, because the source generater is also able to create derived classes.
Also here the generater attaches the `GeneratedCode` attribute again but this does not work since the attribute can not be inherited.

A workaround would be that I check if already the GeneratedCode attribute is there, but is the intended behavior?

**Comment by CZEMacLeod** [comment, marked as answer]
@msedi What exact error are you getting with this scenario?
I did a couple of checks, and whilst you are correct that the attribute cannot be inherited; that is, you cannot create a `MyGeneratedCodeAttribute` that derives from `GeneratedCodeAttribute`, applying the attribute to a class does not prevent it being inherited. 
If you apply the sealed modifier on the class you generate, it would do that.
Also, any class you inherit from a generated class, does not inherit the attribute, such that you can debug, or otherwise work with, manually generated classes that inherit from generated classes.
But a quick test didn't give any compile errors for a derived class and base class both having the attribute explicitly applied to them.

Note that since the GeneratedCodeAttribute has an `AttributeUsageAttribute` that set Inherited to false, even passing true to IsDefined does not return a true for a derived class that does not have the attribute manually applied.

```cs
using System.CodeDom.Compiler;

namespace Test;

[GeneratedCode("Test","1.0.0")]
public class BaseClass
{
}

[GeneratedCode("Test", "1.0.0")]
public class DerivedClass : BaseClass
{
}

public class ManuallyDerivedClass : BaseClass
{
}

public static class Program
{
	public static void Main()
	{
		Type bc = typeof(BaseClass);
		Type dc = typeof(DerivedClass);
		Type mdc = typeof(ManuallyDerivedClass);

		Type gc = typeof(GeneratedCodeAttribute);

		Console.WriteLine($"BaseClass is generated: {bc.IsDefined(gc,false)}");
		Console.WriteLine($"DerivedClass is generated: {dc.IsDefined(gc, false)}");
		Console.WriteLine($"ManuallyDerivedClass is generated: {mdc.IsDefined(gc, false)}");

		Console.WriteLine($"BaseClass is generated: {bc.IsDefined(gc, true)}");
		Console.WriteLine($"DerivedClass is generated: {dc.IsDefined(gc, true)}");
		Console.WriteLine($"ManuallyDerivedClass is generated: {mdc.IsDefined(gc, true)}");
	}
}
```
```
BaseClass is generated: True
DerivedClass is generated: True
ManuallyDerivedClass is generated: False
BaseClass is generated: True
DerivedClass is generated: True
ManuallyDerivedClass is generated: False
```

**Reply by msedi** [reply]
@CZEMacLeod: Thanks for coming back on this. I think I put a wrong title in it. I know that I cannot derive the GeneratedCodeAttribute but as you already recognized that is not what I wanted. Let me check if I did a mistake, because I did some additional tests and there it worked. So something must be wrong on my side, I will check again and come back.

**Reply by CZEMacLeod** [reply]
No problem. It is always worth doing a quick sanity check on a problem to see if it is actually the problem you thought it was. Sometimes just trying to write down the issue will help you find out what it was.
One thing to check is if your base class is sealed - maybe that is what you were having an issue with?

**Reply by msedi** [reply]
@CZEMacLeod: So I made a mistake, Obviously I forgot to restart Visual Studio and this mixed up things. It happened while switching through many branches. Nevertheless, thanks for your feedback.
